### ISSUE TITLE
Feature Request - Built in attachment viewer

### ISSUE THREAD

**Issue by Zangoran** [Foundry376/Mailspring#234]
**Feature Request**
##### Does this feature exist in another mail client or tool you use?
In GMail, when you open a pdf, image, docx, etc. It will open a preview of it inside the email. This would be a great addition to Mailspring, as it would help streamline workflow. Thanks for working on this even when Nylas dropped it.

**Comment by Zangoran** [comment]
@Foundry376 @bengotow Not sure if this was seen yet.

**Comment by bengotow** [comment]
I think this would be great–on Mac OS X you can preview most common attachments directly within the app via Quicklook (by clicking the "eye" that appears on the attachments, which also get larger thumbnail previews). It'd be great to do something like this on Windows, Linux if anyone wants to dive in. It might be tricky though because we don't want the app to try to load PDFs, etc. in the same process that is driving the UI, since a large PDF or 10MP image could "lock" the app for seconds.

<img width="336" alt="image" src="https://user-images.githubusercontent.com/1037212/33512500-a1e87018-d6e6-11e7-9f0f-26b30a94bef9.png">

**Comment by Zangoran** [comment]
@bengotow Thanks for making this a little more known! I hope someone eventually would have to drive to make this down the road. It would help keep everything in the same app.

**Comment by cmstactics** [comment]
Love Mailspring so far and this is one feature that is definitely needed and I may jump back to Thunderbird due to this.  Also displaying images inline instead of attachments would be helpful.

**Comment by leigh-1** [comment]
Just adding my agreement with the preview suggestion - the only reason I have to stick with Office/ Outlook (and now looking at installing Crossover on Linux to install my MS Office on Linux) is due to the attachment previewer..  if they all do the same, there is no reason to use one over the other, and they do considerably less than Outlook without it! That's what makes the difference.. I never understand why software vendors don't put it priority 1 (after the basic send and receive which of course is a given) you can just click down your emails, and if there is an attachment it previews it - eg CVs / documents - we all receive 200 emails per day, so to open an email, select an attachment, open it.. view it, expand it, close it, close the email (5-6 seconds) when you can just hit 'down arrow' and scan down your emails.. and previews attachments without having to do anything (1 second) = 500-600% increased efficiency and saves 20 mins per day ===a WHOLE DAY per month just sat opening attachments, reviewing, closing.. grr! If anyone other than Microsoft would just build a desktop app with a good attachment previewer...   in fact I think I might try and build one... ;-)

**Comment by ericnewton76** [comment]
I agree, would be great to easily preview small attachments.  

Is there any existing "previewer" type libraries out there for nodejs/electron that can render either into html or an html5 canvas?  

I might try to work on that, building an interface that the previewer library can use to render the attachments by providing it with a byte stream somehow and expecting an html result that might be html5 canvas+possibly some browser js.

**Comment by tankerkiller125** [comment]
@ericnewton76 https://www.npmjs.com/package/filepreview has a very large collection of the file is can convert, unfortunately however it renders them into GIF files and not HTML or a HTML Canvas. With that said I think a good starting point would be a PDF viewer and an image viewer as many document types can be converted to PDFs

**Comment by bengotow** [comment]
Hmm `filepreview` looks interesting. Unfortunately I think depending on `ffmpeg`, `imagemagick`, etc. is super practical from a server perspective but would be difficult to package up in a distributable app. I'm sure at least one of those needs to be installed with `sudo` and dumps stuff all over the place...

One option would be to allow users to opt-in to some sort of server-rendered previews, but we'd need to send attachments to the cloud. I can't imagine many folks would like that, especially if it was automatic.

I wonder how many formats we could get Chromium / Electron itself to preview? We might be able to open an offscreen window and capture it's contents. I know it can't open word docs, but it may be able to preview PDFs. If we could get "Office" formats converted to PDF or HTML I think we'd have office+images+PDFs which would be a good start.

**Comment by bengotow** [comment]
Hey folks! I worked on this a bunch today and got a proof of concept going on a branch here: https://github.com/Foundry376/Mailspring/tree/bengotow/quicklook-js.

It uses SheetsJS, PDFJS, Mammoth (for docx) and snarkdown (for markdown) and implements a replacement for both the quicklook thumbnails and the in-app "preview" window that opens when you  click the eye icon!

I need to add a place in the app where we can show the licenses for these librarys because a few of them are Apache 2, but otherwise I think this will be a good start! If you want to add support for more formats I think this would be a great area to contribute. I tried to find a Powerpoint -> HTML/PNG module and couldn't find one that didn't rely on you having all of OpenOffice installed.

I also reached out to the SheetsJS author because he has a "pro" version of the module that requires a paid license but does a lot more. Going to see how much it costs, maybe it's not too much and we can get super nice excel previews.
